src/custom/pages/Profile/AddressSelector.tsx

```diff
@@ -0,0 +1,213 @@
+import { useCallback, useEffect, useRef, useState } from 'react'
+import styled, { css } from 'styled-components/macro'
+import { Check, ChevronDown } from 'react-feather'
+import { useOnClickOutside } from 'hooks/useOnClickOutside'
+import { useActiveWeb3React } from 'hooks/web3'
+import { ensNames } from './ens'
+import { useAddress } from 'state/affiliate/hooks'
+import { updateAddress } from 'state/affiliate/actions'
+import { useAppDispatch } from 'state/hooks'
+import { isAddress, shortenAddress } from 'utils'
+
+type AddressSelectorProps = {
+  address: string
+}
+
+export default function AddressSelector(props: AddressSelectorProps) {
+  const { address } = props
+  const dispatch = useAppDispatch()
+  const selectedAddress = useAddress()
+  const { chainId, library } = useActiveWeb3React()
+  const [open, setOpen] = useState(false)
+  const [items, setItems] = useState<string[]>([address])
+  const toggle = useCallback(() => setOpen((open) => !open), [])
+  const node = useRef<HTMLDivElement>(null)
+  useOnClickOutside(node, open ? toggle : undefined)
+
+  const handleSelectItem = useCallback(
+    (item: string) => {
+      dispatch(updateAddress(item))
+      toggle()
+    },
+    [dispatch, toggle]
+  )
+
+  useEffect(() => {
+    if (!chainId) {
+      return
+    }
+
+    ensNames(chainId, address).then((response) => {
+      if ('error' in response) {
+        console.info(response.error)
+        setItems([address])
+        return
+      }
+      setItems([...response, address])
+    })
+  }, [address, chainId])
+
+  useEffect(() => {
+    if (selectedAddress) {
+      return
+    }
+
+    dispatch(updateAddress(address))
+  }, [selectedAddress, address, dispatch])
+
+  useEffect(() => {
+    if (!selectedAddress) {
+      return
+    }
+
+    // if the user switches accounts, reset the selected address
+    if (isAddress(selectedAddress) && selectedAddress !== address) {
+      dispatch(updateAddress(address))
+      return
+    }
+
+    // the selected address is a ens name, verify that resolves to the correct address
+    const verify = async () => {
+      const resolvedAddress = await library?.resolveName(selectedAddress)
+      if (resolvedAddress !== address) {
+        dispatch(updateAddress(address))
+      }
+    }
+
+    verify()
+  }, [selectedAddress, address, dispatch, library])
```

Comment:
These two `useEffect` seem to depend on almost the same deps. Does it make sense to merge them into a single one?